### ISSUE TITLE
Repeated blocks inside a resource do not map to a list variable

### ISSUE THREAD

**Issue by ericlagergren** [hashicorp/terraform#15911]
### Terraform Version

0.10.2

### Terraform Configuration Files

N/A

### Debug Output

https://gist.github.com/ericlagergren/redacted

### Crash Output

N/A

### Expected Behavior

`terraform plan` to successfully run.

### Actual Behavior

```
1 error(s) occurred:

* module.b.var.list_of_maps: variable list_of_maps in module b should be type list, got map
```

### Steps to Reproduce

```
0:0 /tmp/a $ tree
.
├── a.tf
└── b
    └── b.tf

1 directory, 2 files
0:0 /tmp/a $ cat a.tf
provider "aws" {
  region = "us-west-2"
}

module "b" {
  source = "b/"

  list_of_maps {
    name = "name1"
  }

  list_of_maps {
    name = "name2"
  }

  list_of_maps {
    name = "name3"
  }
}
0:0 /tmp/a $ cat b/b.tf
variable "list_of_maps" {
  type = "list"
}

resource "aws_ecs_cluster" "c" {
  count = "${length(var.list_of_maps)}"
  name  = "${lookup(element(var.list_of_maps, count.index), "name")}"
}
0:0 /tmp/a $ terraform plan
1 error(s) occurred:

* module.b.var.list_of_maps: variable list_of_maps in module b should be type list, got map
```

### Important Factoids

N/A

### References
- https://github.com/hashicorp/terraform/issues/7142#issuecomment-225449953

**Comment by apparentlymart** [comment]
Hi @ericlagergren! Sorry this doesn't work properly.

In spite of what was claimed in the comment you linked, this sort of usage is not really officially supported, due to limitations of the configuration language parser.

With that said, we do intend to make this sort of configuration possible in future. We're currently designing a set of improvements to the configuration language which include making make lists of maps (and other complex structures) work more robustly.

This will, however, use a list assignment syntax rather than a multiple-block syntax, due to that mapping better onto the idea of a variable being a single list value, rather than a set of separate objects:

```hcl
# NOT YET SUPPORTED
# (may change slightly before final implementation)

module "b" {
  source = "b/"

  list_of_maps = [
    {
      name = "name1"
    },
    {
      name = "name2"
    },
    {
      name = "name3"
    },
  ]
}
```

At this time, we recommend usage of only flat structures (strings, lists of strings, maps of strings) as much as possible. Although deeper structures are partially supported, there are various limitations and caveats including the syntax limitation you found here, for historical reasons.

In today's implementation it _is_ possible to construct such values, but it requires some inconvenient syntax via the interpolation language, to work around the limitations of the structural parser by building the structure dynamically:

```hcl
module "b" {
  source = "b/"

  list_of_maps = "${list(map("name", "name1"), map("name", "name2"), map("name", "name3"))}"
}
```

Making something like my first example above work is already in our plans, but I'll leave this issue open here to represent the request and post more updates when we get closer to implementation.

**Comment by rkr-kununu** [comment]
Just to complete this example...  It was also suggested that you don't mix `lookup(element(...` instead use the index operator `[]` and then `lookup()`.  See #8047

```
resource "aws_ecs_cluster" "c" {
  count = "${length(var.list_of_maps)}"
  name = "${lookup(var.list_of_maps[count.index], "name")}"
}
```

If you make this change and the one suggested above, then your example will work with TF 0.10.3:

```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

  + module.b.aws_ecs_cluster.c[0]
      name: "name1"

  + module.b.aws_ecs_cluster.c[1]
      name: "name2"

  + module.b.aws_ecs_cluster.c[2]
      name: "name3"


Plan: 3 to add, 0 to change, 0 to destroy.
```

**Comment by f0o** [comment]
@rkr-kununu This solution will force all resources to be renewed when you append another map to the list though :/

**Comment by apparentlymart** [comment]
Hi all!

This issue is covering similar ground as #3388, which I closed today due to it being addressed by some new features in the v0.12.0 release. I left [a comment over there](https://github.com/hashicorp/terraform/issues/3388#issuecomment-433497150) showing an example I used to verify this before closing it.

Since there's a solution in `master` ready to release, I'm going to close this out now. Thanks for reporting it, and thanks for your patience while we did the groundwork to make this possible.
